include/rmm/mr/device/arena_memory_resource.hpp

```diff
@@ -0,0 +1,293 @@
+      }
+    } else {
+      for (auto& kv : stream_arenas_) {
+        // Check the per-stream arena if it does not belong to the current stream, and return if the
+        // pointer is found.
+        if (kv.first != stream && kv.second.deallocate(p, bytes)) return;
+      }
+    }
+
+    // Allocation not found.
+    RMM_LOGGING_ASSERT(0);
+  }
+
+  /**
+   * @brief Get the arena associated with the current thread or the given stream.
+   *
+   * @param stream The stream associated with the arena.
+   * @return arena& The arena associated with the current thread or the given stream.
+   */
+  arena& get_arena(cudaStream_t stream)
+  {
+    if (use_per_thread_arena(stream)) {
+      return get_thread_arena();
+    } else {
+      return get_stream_arena(stream);
+    }
+  }
+
+  /**
+   * @brief Get the arena associated with the current thread.
+   *
+   * @return arena& The arena associated with the current thread.
+   */
+  arena& get_thread_arena()
+  {
+    auto const id = std::this_thread::get_id();
+    {
+      read_lock lock(mtx_);
+      auto const it = thread_arenas_.find(id);
+      if (it != thread_arenas_.end()) { return it->second; }
+    }
+    {
+      write_lock lock(mtx_);
+      thread_arenas_.emplace(id, global_arena_);
+      return thread_arenas_.at(id);
+    }
+  }
+
+  /**
+   * @brief Get the arena associated with the given stream.
+   *
+   * @return arena& The arena associated with the given stream.
+   */
+  arena& get_stream_arena(cudaStream_t stream)
+  {
+    RMM_LOGGING_ASSERT(!use_per_thread_arena(stream));
+    {
+      read_lock lock(mtx_);
+      auto const it = stream_arenas_.find(stream);
+      if (it != stream_arenas_.end()) { return it->second; }
+    }
+    {
+      write_lock lock(mtx_);
+      stream_arenas_.emplace(stream, global_arena_);
+      return stream_arenas_.at(stream);
+    }
+  }
+
+  /**
+   * @brief Get free and available memory for memory resource.
+   *
+   * @param stream to execute on.
+   * @return std::pair containing free_size and total_size of memory.
+   */
+  std::pair<std::size_t, std::size_t> do_get_mem_info(cudaStream_t stream) const override
+  {
+    return std::make_pair(0, 0);
+  }
+
+  /**
+   * @brief Should a per-thread arena be used given the CUDA stream.
+   *
+   * @param stream to check.
+   * @return true if per-thread arena should be used, false otherwise.
+   */
+  static bool use_per_thread_arena(cudaStream_t stream)
+  {
+#ifdef CUDA_API_PER_THREAD_DEFAULT_STREAM
+    return stream == cudaStreamDefault || stream == cudaStreamPerThread;
+#else
+    return stream == cudaStreamPerThread;
+#endif
+  }
+
+  /// The global arena to allocate superblocks from.
+  global_arena global_arena_;
+  /// Arenas for default streams, one per thread.
+  std::unordered_map<std::thread::id, arena> thread_arenas_;
+  /// Arenas for non-default streams, one per stream.
+  std::unordered_map<cudaStream_t, arena> stream_arenas_;
```

Comment:
I believe @harrism has found for small numbers of elements that `std::map` is faster than `unordered_map`. Furthermore, using a `std::map` would allow a binary search in `deallocate_from_other_arena`.

Reply:
I think the most common use for these maps is to do lookup based on the key, which is O(1) for `unordered_map` and O(log(n)) for `map`, but maybe that doesn't matter if `n` is small.

We do need to iterate through the whole map in `deallocate_from_other_arena` since we don't know where the block was originally allocated.

Reply:
> I think the most common use for these maps is to do lookup based on the key, which is O(1) for unordered_map and O(log(n)) for map, but maybe that doesn't matter if n is small.

Yes, but `unordered_map` has a much larger constant factor than `map` that when `n` is small `std::map` is actually better.

Reply:
https://quick-bench.com/q/wO1w2DLaVJXH08gngu03oNcsOj0

Reply:
Done.

Reply:
https://gist.github.com/harrism/18879bf001b976a443fe0dd319156da6